ion/reader.go

```diff
@@ -85,7 +85,7 @@ type Reader interface {
 	// FieldName returns the field name associated with the current value. It returns
 	// the empty string if there is no current value or the current value has no field
 	// name.
-	FieldName() string
+	FieldName() *string
```

Comment:
Returning a `*string` via the public interface feels pretty awkward an un-go-like to me, especially given that it's going to be non-nil in basically all cases where someone might call it. :( As I suggested previously, I think the interface would be better with this left as a string, and an auxiliary `HasFieldName() bool` method for the rare case where someone is uncertain about whether an empty-string return value means "not in a struct" or "whoever wrote this struct is rude." Thoughts?

If you are going to proceed with this change despite my objections, which is your right, the comment should be updated to match. :)

Reply:
> ...and an auxiliary HasFieldName() bool method...

From what I can tell, `HasFieldName()` is essentially a synonym for `IsInStruct()`. Is that right?

> Returning a *string via the public interface feels pretty awkward an un-go-like to me

Could you elaborate on this? I liked the idea of returning a `*string` because it makes it very clear that there might not be a field name.

This:
```go
if reader.IsInStruct() {
   fieldName := *reader.FieldName()
}
```
feels very similar to this:
```go
if reader.HasFieldName() {
  fieldName := reader.FieldName()
}
```

On the other hand, detecting an absent field name after-the-fact is simpler with a pointer. Compare this:
```go
fieldName := reader.FieldName()
if fieldName == "" && !reader.HasFieldName() {
  // There wasn't a field name
}
```

with this:
```go
fieldName := reader.FieldName()
if fieldName == nil {
  // There wasn't a field name
}
```

Using a fieldName that's `nil` will lead to an error instead of silently proceeding with a surprising value. Using a `string` does let you get away with simply calling `FieldName()` whether it's sensible or not, but I'm unsure that's something we want to encourage.

Reply:
Pedantically speaking, `HasFieldName()` is `IsInStruct() && Type() != NoType`; it will return false immediately after a call to `StepIn()`, for example. :)

Pointers to basic types like strings feel weird because they don't really get used much in idiomatic go (e.g. see all the complaints [here](https://github.com/aws/aws-sdk-go-v2/issues/205) and in linked issues, although in that case I'm not sure there's a better solution). In particular: syntax sugar for pointers-to-structs means it's very uncommon to ever have to manually dereference a pointer; doing so feels strange and a bit dangerous, the kind of thing you're only supposed to do if you're interfacing with a C library.

If you think it's common for folks to check if `FieldName()` has in fact returned a field name, a more idiomatic approach would be `FieldName() (string, bool)`, in the style of go built-in [map type](https://blog.golang.org/maps). My concern with that approach here is that, in almost all cases, I expect the caller already knows to expect a field name by virtue of being inside of a struct (e.g. all three references to FieldName in `unmarshal.go`). It feels unfortunately to make them all explicitly ignore the boolean that they know at compile time will always be true.

```go
// can't do write(in.FieldName())
fn, _ := in.FieldName()
write(fn)
```

In the relatively niche case where someone's handling a sequence of values that may or may not have field names at runtime (as in [process.go](https://github.com/amzn/ion-go/blob/master/cmd/ion-go/process.go#L178)), it seems pretty clear to do:

```go
if in.HasFieldName() {
  out.FieldName(in.FieldName())
}
```

vs

```go
fn, ok := in.FieldName()
if ok {
  out.FieldName(fn)
}
```

(In practice, I think it's probably fine if you feel differently, I imagine most folks who are concerned about idiomatic go will be using Unmarshal not the underlying Reader. :))

Reply:
A field name that is the empty string `""` is *valid* per the Ion Spec. http://amzn.github.io/ion-docs/docs/spec.html#struct. 

So there is a case where we can be inside a struct and the empty string is a valid value for a field name. So there is going to be a test in the client code to check what `""` means.  This means that whether we move to `string` or `*string` there is still a test that needs to be performed by the client if they are to validate correct behaviour here. 

I also agree with the comment that "this is the low level API" and as such there is an expectation that the client code needs to know of these cases. 

I vote for a 1-1 mapping of value to case here 

* `nil` -- something is wrong go figure out what by checking the reader's context 
* non-nil -- all is well you can proceed 

We should make this **explicit** and explain it in the comment for the method. 

Another argument here (though maybe weak) is that in cases where the API is misused and the client does not check the value, a `nil` will error while an `""` can result in data that is valid Ion but not the intended Ion and detecting that you are generating valid Ion but invalid for the use case is harder and riskier in my opinion. 

My vote is to keep the pointer type and add a comment to the API that explicitly spells out the distinction between `nil` and non-`nil` values so that clients are aware of the methods expectations and guarantees.